[OOPOOD-2021-fall] hw5

IgorOzherelev opened the pull request.

Commit:
refactor-completed

Commit:
small fixes

Commit:
hometask-2 added

Commit:
AbstractMovableGraphicObject->AbstractMovableObject no longer extends from GraphicObject

Commit:
добавлены тесты

Commit:
тесты на танк перемещены

Commit:
added logic map generators independent from gdx

Commit:
added random and file map generation

Commit:
Merge remote-tracking branch 'origin/hometask-3' into hometask-3

# Conflicts:
#	src/main/java/ru/mipt/bit/platformer/GameDesktopListener.java
#	src/main/java/ru/mipt/bit/platformer/models/movable/AbstractMovableObject.java
#	src/main/java/ru/mipt/bit/platformer/models/movable/Tank.java
#	src/main/java/ru/mipt/bit/platformer/models/objects/LibGdxGraphicObject.java
#	src/main/java/ru/mipt/bit/platformer/services/generator/MapGenerator.java
#	src/main/java/ru/mipt/bit/platformer/services/renderers/LibGdxGraphicObjectsRendererService.java
#	src/test/java/ru/mipt/bit/platformer/utils/GameGraphicObjectCollisionUtilsTest.java

Commit:
branches conflict resolved

Commit:
added tests + some refactor

Commit:
corrected splitFile method

Commit:
corrected preferences

Commit:
Добавлена реализация AI

Commit:
логика столкновений через CollidingManagerService

Commit:
исходная клетка и соседняя считаются занятыми

Commit:
[After lecture]: some changes for commands, bots + renaming

Commit:
added ai dependency

Commit:
[After lecture]: added custom ai class

Commit:
Merge branch 'hometask-4' into hometask-5

Commit:
implemented support for external ai

Commit:
renaming

a-ivanov commented on src/main/java/ru/mipt/bit/platformer/models/movable/LibGdxTank.java:
```diff
@@ -0,0 +1,34 @@
+import ru.mipt.bit.platformer.models.objects.GameObject;
+
+
+public class LibGdxTank extends AbstractLibGdxMovableObject {
```
У танка не должно быть ничего от LibGdx в названии. Это может быть у графического класса, если он рисуется ее средствами.

IgorOzherelev replied:
Исправлено

a-ivanov commented on src/main/java/ru/mipt/bit/platformer/models/objects/GameObject.java:
```diff
@@ -0,0 +1,55 @@
+
+import java.util.Objects;
+
+public class GameObject implements Colliding {
```
Не стоит переусердствовать при разбиении логики между классами. Достаточно (уже обсуждали на лекциях) по классу на:
- логику (танк перемещается между клетками, проверка коллизий; препятствие проверяется на коллизии)
- графику (танк перемещается по экрану, препятствие рисуется на одном и том же месте).

IgorOzherelev replied:
То есть получается, что движение будет завязано на графику?

a-ivanov replied:
Так нет же, в классе с движением - нет ссылки на графику. В классе с графикой - есть ссылка на движение. Итого графика завязана на логику (движение).

IgorOzherelev replied:
Исправлено

a-ivanov commented on src/main/java/ru/mipt/bit/platformer/models/movable/Movable.java:
```diff
@@ -0,0 +1,15 @@
+
+public interface Movable extends Colliding {
+    float continueProgress(float previousProgress, float deltaTime, float speed);
+    void prepareForMove(Direction direction);
```
Слишком много деталей реализации в публично апи. Им сложно пользоваться. Тут явно многое - методы реализации.

IgorOzherelev replied:
Исправлено

a-ivanov commented on src/main/java/ru/mipt/bit/platformer/services/renderers/LibGdxTreesRendererService.java:
```diff
@@ -0,0 +1,51 @@
+
+import static ru.mipt.bit.platformer.util.GameGraphicUtils.*;
+
+public class LibGdxTreesRendererService implements RendererService {
```
Не надо разбивать все на "сервисы". Достаточно для простоты вызывать релевантные методы в классе Level (уровень делегирует вызов методов tick для своих логических объектов)/LevelRenderer (вызывает сначала отрисовку клеток, потом танков/препятствий).

IgorOzherelev replied:
Исправлено

a-ivanov commented on src/main/java/ru/mipt/bit/platformer/services/renderers/LibGdxTanksRendererService.java:
```diff
@@ -0,0 +1,46 @@
+        this.tanks = createLibGdxTanks(storage.getTanks());
+    }
+
+    public List<? extends Movable> getTanks() {
```
Создавайте геттеры только когда они реально нужны, всегда стоит писать код без них, чтобы логика работы с данным оставалась в private методах класса.

IgorOzherelev replied:
Ну, так он используется в `LibGdxLevelRendererService`

a-ivanov replied:
LibGdxLevelRendererService (переименуйте) должен делегировать отрисовку LibGdxTanksRendererService, не вызывать геттеры из него. Иначе сама логики отрисовки танка - в LibGdxLevelRendererService, а не в LibGdxTanksRendererService.

IgorOzherelev replied:
Исправлено

a-ivanov commented on src/main/java/ru/mipt/bit/platformer/services/gamestate/AwesomeAiGameStateService.java:
```diff
@@ -0,0 +1,57 @@
+import static ru.mipt.bit.platformer.services.converters.awesomeai.MovableBotConverterService.convertMovablesToBots;
+import static ru.mipt.bit.platformer.services.converters.awesomeai.MovablePlayerConverterService.convertMovableToPlayer;
+
+public class AwesomeAiGameStateService implements AiGameStateService<GameState> {
```
Не вижу интерфейс для создания уровня и две реализации: случайное расположение и чтение из файла. Для этого еще должен быть класс для уровня с логическими объектами, чтобы его можно было создавать.

IgorOzherelev replied:
Создание уровня реализовано в пакете generator

a-ivanov replied:
Понял, см. замечания там.

IgorOzherelev replied:
Исправлено

a-ivanov commented on src/main/java/ru/mipt/bit/platformer/services/ai/CustomAiService.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.util.List;
+
+public interface CustomAiService {
```
Нет, по заданию не должно быть явных интерфейсов, говорящих об использовании сторонней библиотеки. У вас должен появиться интерфейс AIController, с реализацией отправки случайных движений ботов. Далее в 5 дз вам необходимо написать адаптер. Не вижу ни того, ни другого.

IgorOzherelev replied:
В этом пакете реализован интерфейс для обычного ai, и вроде бы реализован адаптер для стороннего: `Awesome AICustomService`. В нем содержится поле с внешним ai, сам класс реализует интерфейс `CustomAIService`; разве адаптер может быть только с наследованием, которое по лекциям плохо?

a-ivanov replied:
Понял, да, похоже на правду. CustomAiService - плохое название, это должен быть единственный источник команд для игрока/ботов, тогда так и называйте его, например TankController.

IgorOzherelev replied:
Исправлено

a-ivanov commented on src/main/java/ru/mipt/bit/platformer/services/commands/BotsCommandExecutorService.java:
```diff
@@ -0,0 +1,17 @@
+import ru.mipt.bit.platformer.models.commands.Command;
+import ru.mipt.bit.platformer.services.ai.CustomAiService;
+
+public class BotsCommandExecutorService implements CommandExecutorService {
```
Тут должна быть логика отправки команд ботам. Далее по 5 дз необходимо написать еще одну имплементацию этого интерфейса-адаптер, который будет получать рекомендации, переводить их в команды и отправлять ботам.

IgorOzherelev replied:
Что вы имеете в виду под отправкой команд ботам? Сейчас сделаю так: каждой команде прописывается бот, команды рекомендуется от ai сервисов, здесь они вызываются

a-ivanov replied:
Да, было непонятно с CustomAiService, см. замечания к нему.

IgorOzherelev replied:
Исправлено

a-ivanov commented on src/main/java/ru/mipt/bit/platformer/services/generator/MapGenerator.java:
```diff
@@ -0,0 +1,7 @@
+import ru.mipt.bit.platformer.models.storages.GameObjectStorage;
+
+public interface MapGenerator {
+    GameObjectStorage generate();
```
Возвращайте Level, промежуточные бесполезные классы не нужны.

IgorOzherelev replied:
Исправлено

a-ivanov commented on src/main/java/ru/mipt/bit/platformer/services/ai/CustomAiService.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface CustomAiService {
+    List<? extends Command> recommend();
+    void setDeltaTime(float deltaTime);
```
Если deltatime нужен в recommend, передавайте его туда аргументом, не создавайте мутирующих апи (которые можно забыть вызвать).

IgorOzherelev replied:
Исправлено

Commit:
[After review]:
      [hw2] - исправлено нарушение SRP:
            Добавлены 2 логических и 2 графических класса
            Пакеты: models, graphics, geometry

      [hw3] - исправлено, генераторы возвращают класс Level
            Удалены лишние классы, избавление от сеттеров
            Пакеты: level, utils

      [hw4] - исправлено, логика Ai перенесена на уровень
            контроллеров;
            Управление игроком через команды
            Пакеты: commands, actions, controllers

      [hw5] - исправлено, паттерн адаптер для внешнего Ai
            перенесен на уровень контроллеров.
            Пакеты: controllers, converters

IgorOzherelev commented on src/main/java/ru/mipt/bit/platformer/services/movement/TileMovementService.java:
```diff
@@ -0,0 +1,7 @@
+
+import ru.mipt.bit.platformer.models.movable.Movable;
+
+public interface TileMovementService {
```
Исправлено

IgorOzherelev replied:
Исправлено

IgorOzherelev commented on src/main/java/ru/mipt/bit/platformer/models/movable/AbstractLibGdxMovableObject.java:
```diff
@@ -0,0 +1,102 @@
+
+    protected float movementProgress = MAX_PROGRESS;
+
+    protected LibGdxGraphicObject libGdxGraphicObject;
```
Исправлено

IgorOzherelev replied:
Исправлено

IgorOzherelev commented on src/main/java/ru/mipt/bit/platformer/services/renderers/LibGdxLevelRendererService.java:
```diff
@@ -0,0 +1,74 @@
+import static com.badlogic.gdx.graphics.GL20.GL_COLOR_BUFFER_BIT;
+import static ru.mipt.bit.platformer.util.LibGdxGameLevelUtils.createSingleLayerMapRenderer;
+
+public class LibGdxLevelRendererService implements RendererService {
```
Исправлено
